flask_admin/contrib/fileadmin/s3.py

```diff
@@ -0,0 +1,161 @@
+try:
+    from boto import s3
+    from boto.s3.prefix import Prefix
+    from boto.s3.key import Key
+except ImportError:
+    s3 = None
+
+import dateutil.parser
+
+from flask import redirect
+from flask_admin.babel import gettext
+
+
+class S3Storage(object):
+    """
+        Storage object representing files on an Amazon S3 bucket.
+
+        Usage::
+
+            from flask_admin.contrib.fileadmin import BaseFileAdmin,
+            from flask.admin.contrib.fileadmin.s3 import S3Storage
+
+            class MyS3Admin(BaseFileAdmin):
+                # Configure your class however you like
+                pass
+
+            fileadmin_view = MyS3Admin(storage=S3Storage(...))
+
+    """
+
+    def __init__(self, bucket_name, region, aws_access_key_id,
+                 aws_secret_access_key):
+        """
+            Constructor
+
+                :param bucket_name:
+                    Name of the bucket that the files are on.
+
+                :param region:
+                    Region that the bucket is located
+
+                :param aws_access_key_id:
+                    AWS Access Key ID
+
+                :param aws_secret_access_key:
+                    AWS Secret Access Key
+
+            Make sure the credentials have the correct permissions set up on
+            Amazon or else S3 will return a 403 FORBIDDEN error.
+        """
+
+        if not s3:
+            raise ValueError('Could not import boto. You can install boto by '
+                             'using pip install boto')
+
+        connection = s3.connect_to_region(region,
+                                          aws_access_key_id=aws_access_key_id,
+                                          aws_secret_access_key=
+                                          aws_secret_access_key)
+        self.bucket = connection.get_bucket(bucket_name)
+        self.separator = '/'
+
+    def get_files(self, path, directory):
+        def _strip_path(name, path):
+            if name.startswith(path):
+                return name.replace(path, '', 1)
+            return name
+
+        def _remove_trailing_slash(name):
+            return name[:-1]
+
+        files = []
+        directories = []
+        if path and not path.endswith(self.separator):
+            path += self.separator
+        for key in self.bucket.list(path, self.separator):
+            if key.name == path:
+                continue
+            if isinstance(key, Prefix):
+                name = _remove_trailing_slash(_strip_path(key.name, path))
+                key_name = _remove_trailing_slash(key.name)
+                directories.append((name, key_name, True, 0, 0))
+            else:
+                last_modified = int(dateutil.parser.parse(key.last_modified)
```

Comment:
Here's an example that uses strptime: https://github.com/jordansissel/s3cleaner/blob/5969afaf5edf0223f20e13a21aa190d20c6b958a/s3cleaner.py#L46